GeocentricLatLonAlt

uriele opened the pull request:
Add support for GeocentricLatLonAlt

Add direct conversion  GeocentricLatLon <> Cartesian
Add direct conversion  GeocentricLatLon <> Projection
Add conversion GeocentricLatLonAlt<>LatLonAlt

Add tests for Geocentric data

In many practical cases, the latitude and longitude of ESA satellite data are given in Geocentric instead of Geodesic coordinates. GeocentriLatLonAlt converts the information into LatLon before adding the height and converting it into cartesian.

juliohm commented on src/crs/geographic.jl:
```diff
@@ -215,6 +215,56 @@ Random.rand(rng::Random.AbstractRNG, ::Type{GeocentricLatLon{Datum}}) where {Dat
+GeocentricLatLonAlt{Datum}(lat::D, lon::D, alt::M) where {Datum,D<:Deg,M<:Met} =
+  GeocentricLatLonAlt{Datum,float(D),float(M)}(checklat(lat), fixlon(lon), alt)
+GeocentricLatLonAlt{Datum}(lat::Deg, lon::Deg, alt::Met) where {Datum} =
+    GeocentricLatLonAlt{Datum}(promote(lat, lon)..., alt)
+GeocentricLatLonAlt{Datum}(lat::Deg, lon::Deg, alt::Len) where {Datum} =
+    GeocentricLatLonAlt{Datum}(lat, lon, uconvert(m, alt))
+GeocentricLatLonAlt{Datum}(lat::Rad, lon::Rad, alt::Len) where {Datum} =
+    GeocentricLatLonAlt{Datum}(rad2deg(lat), rad2deg(lon), alt)
+GeocentricLatLonAlt{Datum}(lat::Number, lon::Number, alt::Number) where {Datum} =
+    GeocentricLatLonAlt{Datum}(addunit(lat, °), addunit(lon, °), addunit(alt, m))
```
Please run the JuliaFormatter locally, you can do that by right-clicking on VSCode and choosing to format the document. We use 2 whitespaces for indentation.

juliohm commented on src/crs/geographic.jl:
```diff
@@ -369,6 +441,28 @@ function Base.convert(::Type{LatLon{Datum}}, coords::Cartesian{Datum,3}) where {
 end
 
+function Base.convert(::Type{Cartesian{Datum}}, coords::GeocentricLatLon{Datum}) where {Datum}
+  geocentric_lla = convert(GeocentricLatLonAlt{Datum}, coords);
+  return convert(Cartesian{Datum}, geocentric_lla)
```
```suggestion
  lla = convert(GeocentricLatLonAlt{Datum}, coords)
  return convert(Cartesian{Datum}, lla)
```

juliohm commented on src/crs/geographic.jl:
```diff
@@ -369,6 +441,28 @@ function Base.convert(::Type{LatLon{Datum}}, coords::Cartesian{Datum,3}) where {
+end
+
+function Base.convert(::Type{GeocentricLatLon{Datum}}, coords::Cartesian{Datum,3}) where {Datum}
+  geocentric_lla = convert(GeocentricLatLonAlt{Datum}, coords);
+  return convert(GeocentricLatLon{Datum}, geocentric_lla)
```
we avoid underscores in variable names, and don't use explicit return in the last statement

uriele commented:
Thank you for the review:

I have run the formatter, removed the explicit return, and changed geocentric_lla to lla

In the conversion  GeocentriLatLonAlt <> Carthesian, the internal variable geodesic_coords has been renamed _coords to signify an internal variable different from the input variable coords.

uriele commented:
Thank you for the review:

I have run the formatter, removed the explicit return, and changed geocentri_clla to lla

In the conversion  GeocentriLatLonAlt <> Carthesian, the internal variable geodesic_coords has been renamed _coords to signify it is an internal variable different from the input variable coords

juliohm commented:
@uriele did you have a chance to look into the missing tests?

uriele commented:
> @uriele did you have a chance to look into the missing tests?

Hi @juliohm sorry, which missing tests? I think I doubled the test coverage of latlon and latlonalt

juliohm commented on test/crs/conversions.jl:
```diff
@@ -350,6 +350,10 @@
       @inferred convert(LatLon, c2)
     end
 
+    # conversion LatLonAlt <> GeocentricLatLonAlt is tested in the next section
+    @testset "LatLonAlt <> GeocentricLatLonAlt" begin
+    end
```
and these

uriele replied:
added test, I didn't initially put it cause they relay on the LatLon<>GeocentricLatLon conversion being correct and passing the altitude

juliohm commented on test/crs/conversions.jl:
```diff
@@ -350,6 +350,21 @@
       @inferred convert(LatLon, c2)
     end
 
+    # conversion LatLonAlt <> GeocentricLatLonAlt is tested in the next section
```
```suggestion
```

juliohm commented on test/crs/conversions.jl:
```diff
@@ -398,6 +413,51 @@
         @inferred convert(LatLonAlt, c2)
       end
+
+      # conversion GeocentricLatLonAlt <> Cartesian is tested in the next section
```
```suggestion
```

juliohm reviewed (changes requested):
Thank you @uriele, that is a really nice PR. In this first round of review I am just cleaning unnecessary comments and other minor formatting issues.

juliohm commented on test/crs/constructors.jl:
```diff
@@ -262,6 +262,36 @@
       @test_throws ArgumentError GeocentricLatLon(T(1) * °, T(1) * s)
       @test_throws ArgumentError GeocentricLatLon(T(1) * s, T(1) * s)
     end
+    @testset "GeocentricLatLonAlt" begin
```
```suggestion

    @testset "GeocentricLatLonAlt" begin
```

juliohm commented on test/crs/conversions.jl:
```diff
@@ -807,6 +865,18 @@
       @inferred convert(LatLon{WGS84{0}}, c5)
     end
+
+    @testset "LatLonAlt: Datum conversion" begin
+      # TODO
+    end
+
+    @testset "GeocentricLatLon: Datum conversion" begin
+      # TODO
+    end
+
+    @testset "GeocentricLatLonAlt: Datum conversion" begin
+      # TODO
+    end
```
@uriele what do you think of adding tests for these cases for completeness? We will look into the added conversion methods more carefully during the day.

eliascarv commented on src/crs/geographic.jl:
```diff
@@ -422,6 +513,14 @@ end
+  cartₜ = convert(Cartesian{Datumₜ}, cartₛ)
+  convert(GeocentricLatLon{Datumₜ}, cartₜ)
+end
+# avoid converting coordinates with the same datum as the first argument
+Base.convert(::Type{GeocentricLatLon{Datum}}, coords::GeocentricLatLon{Datum}) where {Datum} = coords
```
Yes, this method is needed to avoid calling the method above this one when the `Datum` is the same.

eliascarv commented on src/crs/geographic.jl:
```diff
@@ -422,6 +508,14 @@ end
+  cartₛ = convert(Cartesian{Datumₛ}, coords)
+  cartₜ = convert(Cartesian{Datumₜ}, cartₛ)
+  convert(GeocentricLatLon{Datumₜ}, cartₜ)
+end
+# avoid converting coordinates with the same datum as the first argument
+Base.convert(::Type{GeocentricLatLon{Datum}}, coords::GeocentricLatLon{Datum}) where {Datum} = coords
```
```suggestion
end

# avoid converting coordinates with the same datum as the first argument
Base.convert(::Type{GeocentricLatLon{Datum}}, coords::GeocentricLatLon{Datum}) where {Datum} = coords
```

eliascarv commented on src/crs/projected.jl:
```diff
@@ -116,6 +116,30 @@ include("projected/transversemercator.jl")
 # FALLBACKS
 # ----------
 
+function Base.convert(::Type{C}, coords::GeocentricLatLon{Datum}) where {Datum,C<:Projected{Datum}}
+  convert(C, convert(LatLon{Datum}, coords))
+end
+
+function Base.convert(::Type{GeocentricLatLon{Datum}}, coords::C) where {Datum,C<:Projected{Datum}}
+  convert(GeocentricLatLon{Datum}, convert(LatLon{Datum}, coords))
+end
+
+function Base.convert(::Type{C}, coords::GeocentricLatLonAlt{Datum}) where {Datum,C<:Projected{Datum}}
+  convert(C, convert(GeocentricLatLon{Datum}, coords))
+end
+
+function Base.convert(::Type{GeocentricLatLonAlt{Datum}}, coords::C) where {Datum,C<:Projected{Datum}}
+  convert(GeocentricLatLonAlt{Datum}, convert(GeocentricLatLon{Datum}, coords))
+end
+
+function Base.convert(::Type{C}, coords::LatLonAlt{Datum}) where {Datum,C<:Projected{Datum}}
+  convert(C, convert(LatLon{Datum}, coords))
+end
+
+function Base.convert(::Type{LatLonAlt{Datum}}, coords::C) where {Datum,C<:Projected{Datum}}
+  convert(LatLonAlt{Datum}, convert(LatLon{Datum}, coords))
+end
```
Please use one-line function definition: `f(x) = body`

eliascarv commented on test/crs/conversions.jl:
```diff
@@ -807,6 +865,18 @@
       @inferred convert(LatLon{WGS84{0}}, c5)
     end
+
+    @testset "LatLonAlt: Datum conversion" begin
+      # TODO
+    end
+
+    @testset "GeocentricLatLon: Datum conversion" begin
+      # TODO
+    end
+
+    @testset "GeocentricLatLonAlt: Datum conversion" begin
+      # TODO
+    end
```
I think this PR only adds Datum conversions for `GeocentricLatLon`. `La LonAlt` and `GeocentricLatLonAlt` do not yet have these conversions.

eliascarv replied:
If you want, you can add these methods in this RP.

uriele replied:
I would move them to the next PR, honestly, mostly because I would need to double-check the math by hand or use the output of a conversion as my test, which would probably make the test not particularly significative (like 1=1)

juliohm commented on src/crs/projected.jl:
```diff
@@ -116,6 +116,19 @@ include("projected/transversemercator.jl")
+
+Base.convert(::Type{LatLonAlt{Datum}}, coords::C) where {Datum,C<:Projected{Datum}} = convert(LatLonAlt{Datum}, convert(LatLon{Datum}, coords))
+
+
```
```suggestion
```

juliohm reviewed (approved):
This looks wonderful! ❤️ 

Appreciate the attention to detail. Thank you, @uriele !

juliohm commented:
Available in https://github.com/JuliaRegistries/General/pull/115516
